### PR TITLE
Aspects

### PR DESCRIPTION
**Pull request by flomonster**

### CODE REVIEW

File: examples/tiny_infra/infra.json [hunk 1]
```diff
@@ -24,7 +53,6 @@
   "routes": [
     {
       "id": "rt.buffer_stop_a-C1",
-      "transit_type": "FLEXIBLE",
```
Comment (multun) [inline examples/tiny_infra/infra.json:27]: is it a related change?

File: examples/tiny_infra/config_railjson.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
   "infra_path": "infra.json",
   "simulation_path": "simulation.json",
   "show_viewer": true,
-  "realtime_viewer": true,
+  "realtime_viewer": false,
```
Comment (multun) [inline examples/tiny_infra/config_railjson.json]: that looks unintended

File: src/main/java/fr/sncf/osrd/train/InteractionsType.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package fr.sncf.osrd.train;
+
+import fr.sncf.osrd.utils.SortedArraySet;
+
+import java.util.Arrays;
+
+public class InteractionsType extends SortedArraySet<InteractionType> {
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/train/InteractionsType.java]: how about InteractionTypeSet

File: src/main/java/fr/sncf/osrd/train/TrainState.java [hunk 4]
```diff
@@ -197,35 +196,55 @@ private void step(
                 throw new SimulationError("train physics numerical integration doesn't seem to stop");
             var distanceStep = goalPathPosition - location.getPathPosition();
             step(locationChange, 1.0, distanceStep);
+            if (speed == 0.)
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/train/TrainState.java]: meh

File: src/main/java/fr/sncf/osrd/train/events/TrainMove.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import fr.sncf.osrd.simulation.Simulation;
+import fr.sncf.osrd.simulation.SimulationError;
+import fr.sncf.osrd.simulation.TimelineEvent;
+import fr.sncf.osrd.simulation.TimelineEventId;
+import fr.sncf.osrd.train.Interaction;
+import fr.sncf.osrd.train.Train;
+
+public final class TrainMove extends TimelineEvent {
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/train/events/TrainMove.java]: how about TrainMoveEvent? can you document this class to include the rationale for having two events for the interaction with action points / regular moves?

File: src/main/java/fr/sncf/osrd/train/events/TrainMove.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    public final Train train;
+    public final Train.TrainStateChange trainStateChange;
+
+    /** Event that represents a train's interaction with an action point */
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/train/events/TrainMove.java]: that's not true, please either remove it or change it

**Review by multun — CHANGES_REQUESTED** [review]

File: src/main/java/fr/sncf/osrd/DebugViewer.java [hunk 7]
```diff
@@ -263,9 +264,13 @@ private void updateTrain(TrainData trainData) {
         var nextMove = trainData.nextMove;
 
         var lastUpdate = nextMove.findLastSpeedUpdate(currentTime);
+        if (lastUpdate == null)
+            return;
         var pathHeadPosition = lastUpdate.interpolatePosition(currentTime);
         var pathTailPosition = Double.max(0, pathHeadPosition - trainData.schedule.rollingStock.length);
         var headLocation = trainData.schedule.findLocation(pathHeadPosition);
+        if (headLocation == null)
+            return;
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/DebugViewer.java:273]: This seems like an error case, do we want to ignore it? Or is it an usual state?

File: src/main/java/fr/sncf/osrd/infra/signaling/AspectConstraint.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            this.element = element;
+        }
+
+        /** Convert to a path position given a phase a train state */
+        public double convert(SignalNavigatePhase.State navigatePhase, TrainState trainState) {
+            Predicate<Interaction> isSignal = interaction -> interaction.interactionType == InteractionType.HEAD
+                    && interaction.actionPoint.getClass() == Signal.class;
+            Interaction interactionElement = null;
+            if (element == Element.NEXT_SIGNAL) {
+                AtomicBoolean isNext = new AtomicBoolean(false);
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/infra/signaling/AspectConstraint.java:35]: Why do we need it to be atomic?

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java [hunk 9]
```diff
@@ -133,10 +130,11 @@ private static double computePositionDelta(double currentSpeed, double accelerat
 
     private static double computeTimeDelta(double currentSpeed, double acceleration, double positionDelta) {
         // Solve: acceleration / 2 * t^2 + currentSpeed * t - positionDelta = 0
-        if (acceleration == 0)
+        if (acceleration < 0.00001)
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java:133]: Are we sure acceleration is never negative?

File: src/main/java/fr/sncf/osrd/train/TrainState.java [hunk 10]
```diff
@@ -197,35 +196,56 @@ private void step(
...
         }
 
         return locationChange;
     }
 
+    /**  Create a location change from the current state to current simulation time */
+    public Train.TrainStateChange evolveStateUntilNow(Simulation sim) {
+        var locationChange = new Train.TrainStateChange(sim, trainSchedule.trainID, this);
+
+        while (this.time + 1.0 < sim.getTime())
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/train/TrainState.java:211]: Where does the value `1.0` come from?

Comment (flomonster) [reply src/main/java/fr/sncf/osrd/train/TrainState.java:211]: For the moment we use a fixed time step of 1 second. Ideally we should have a variable time step.

File: src/main/java/fr/sncf/osrd/infra/signaling/AspectConstraint.java (hunk 8 above)
Comment (flomonster) [reply src/main/java/fr/sncf/osrd/infra/signaling/AspectConstraint.java:35]: To capture a variable it needs to be either an atomic or final.

File: src/main/java/fr/sncf/osrd/DebugViewer.java (hunk 7 above)
Comment (flomonster) [reply src/main/java/fr/sncf/osrd/DebugViewer.java:273]: This case occurs when the train finishes its phase with a non-zero speed.

**Pull request merged**